app/lifecycle/hook.go

```diff
@@ -0,0 +1,201 @@
+	fn(ctx)
+	return nil
+}
+
+// HookStartType defines the type of start hook.
+type HookStartType int
+
+const (
+	// AsyncAppCtx defines a start hook that will be called asynchronously (non-blocking)
+	// with the application context. Using the application usually results in hard shutdown.
+	AsyncAppCtx HookStartType = iota + 1
+
+	// SyncBackground defines a start hook that wil be called synchronously (blocking)
+	// with a fresh background context. Processes that support graceful shutdown can
+	// associate this with a call to RegisterStop.
+	SyncBackground
+
+	// AsyncBackground defines a start hook that wil be called asynchronously (non-blocking)
+	// with a fresh background context. Processes that support graceful shutdown can
+	// associate this with a call to RegisterStop.
+	AsyncBackground
+)
+
+// hook represents a life cycle hook; either a start or a stop.
+type hook struct {
+	// Order defines the order in which hooks are called.
+	Order int
+	// Label is a text label for errors and logging.
+	Label string
+	// StartType defines whether the start type is (a)synchronous and which context to use.
+	StartType HookStartType
+	// Func is the hook function.
+	Func IHookFunc
+}
+
+// runHooks starts and stops all the provided hooks.
+func runHooks(appCtx context.Context, startHooks []hook, stopHooks []hook) error {
+	// Collect any first error, to return at the end.
+	firstErr := make(chan error, 1)
+	cacheErr := func(err error) {
+		select {
+		case firstErr <- err:
+		default:
+			// Some other error already first.
+		}
+	}
+
+	// startAppCtx is cancelled when app is shutdown or when starting a hook fails.
+	startAppCtx, cancel := context.WithCancel(appCtx)
+	defer cancel()
+
+	// backgroundCtx is never closed, it is provided to StartSyncBackground and StartAsyncBackground hooks,
+	// they are explicitly stopped.
+	backgroundCtx := log.WithTopic(context.Background(), "app-start")
+
+	err := startAllHooks(startAppCtx, backgroundCtx, startHooks, cancel, cacheErr)
+	if err != nil {
+		return err
+	}
+
+	// Wait for shutdown or hook start failure
+	<-startAppCtx.Done()
+
+	if appCtx.Err() != nil {
+		log.Info(appCtx, "Shutdown signal detected")
+	}
+
+	// stopCtx is a fresh context allowing 10s for shutdown.
+	stopCtx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	stopCtx = log.WithTopic(stopCtx, "app-stop")
+	log.Info(stopCtx, "Shutting down gracefully")
+
+	stopAllHooks(stopCtx, stopHooks, cancel, cacheErr)
+
+	cacheErr(nil) // Ensure there is something in firstErr.
+
+	return <-firstErr
+}
+
+// startHook starts a hook, block until it returns.
+func startHook(ctx context.Context, hook hook, cancel context.CancelFunc, cacheErr func(err error)) {
+	err := hook.Func.Call(ctx)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		cacheErr(errors.Wrap(err, "start hook", z.Str("hook", hook.Label)))
+		cancel()
+	}
+}
+
+// startAllHooks starts all hooks from hook collection.
+func startAllHooks(
```

Comment:
same here, rather stick to normal pattern of returning an error and do fancy context stuff and error in parent

Reply:
so if you mean to do not call `cacheErr()` and `cancel()` and from inside here, and return an error when something happen and the caller exectute `cacheErr()` and `cancel()` that would change the behaviour, as an error in one of the hooks, does not stop the hooks loop.

Also AsyncBackground and AsyncAppCtx hooks would not be able to return the error

Reply:
yeah, passing two contexts and and a cancel function doesn't feel right. I think we should either do a proper refactor that allows for smaller decoupled functions, or just keep it the way it was.

Reply:
we cannot keep it the way it was, the functions are so long that it doesn't even pass linting